openedx_learning/apps/authoring/publishing/api.py

```diff
@@ -449,13 +465,199 @@ def set_draft_version(
+    inside a bulk_draft_changes_for block, it will not add side-effects for
+    containers, as bulk_draft_changes_for will automatically do that when the
+    block exits.
+    """
+    if set_at is None:
+        set_at = datetime.now(tz=timezone.utc)
+    
+    tx_context = atomic() if create_transaction else nullcontext()
+
+    with tx_context:
+        old_version_id = draft.version_id
+        draft.version_id = publishable_entity_version_pk
+        draft.save()
+
+        learning_package_id = draft.entity.learning_package_id
+
+        # Check to see if we're inside a context manager for an active
+        # DraftChangeLog (i.e. what happens if the caller is using the public
+        # bulk_draft_changes_for() API call), or if we have to make our own.
+        active_change_log = DraftChangeLogContext.get_active_draft_change_log(learning_package_id)
+        change_log = active_change_log or DraftChangeLog.objects.create(
+            learning_package_id=learning_package_id,
+            changed_at=set_at,
+            changed_by_id=set_by,
+        )
+
+        if active_change_log:
+            change = _add_to_existing_draft_change_log(
+                active_change_log,
+                draft.entity_id,
+                old_version_id=old_version_id,
+                new_version_id=publishable_entity_version_pk,
+            )
+            # We explicitly *don't* create container side effects here because
+            # there may be many changes in this DraftChangeLog, some of which
+            # haven't been made yet. It wouldn't make sense to create a side
+            # effect that says, "this Unit changed because this Component in it
+            # changed" if we were changing that same Unit later on in the same
+            # DraftChangeLog, because that new Unit version might not even
+            # include the child Component. So we'll let DraftChangeLogContext
+            # do that work when it exits its context.
+        else:
+            # This means there is no active DraftChangeLog, so we create our own
+            # and add our DraftChangeLogRecord to it. This has the minor
+            # optimization that we don't have to check for an existing
+            # DraftChangeLogRecord, because we know it can't exist yet.
+            change = DraftChangeLogRecord.objects.create(
+                draft_change_log=change_log,
+                entity_id=draft.entity_id,
+                old_version_id=old_version_id,
+                new_version_id=publishable_entity_version_pk,
+            )    
+            _create_container_side_effects_for_draft_change(change)
+
+
+@set_draft_version.register(int)
+def _(
+    publishable_entity_id: int,
+    publishable_entity_version_pk: int | None,
+    /,
+    set_at: datetime | None = None,
+    set_by: int | None = None,  # User.id
+    create_transaction: bool = True,
+) -> None:
+    """
+    Alias for set_draft_version taking PublishableEntity.id instead of a Draft.
+    """
```

Comment:
Just wanted to call out that this is the first time I'm introducing this pattern into openedx-learning (or really anywhere in Open edX code that I'm aware of), where we use `functools.singledispatch` to give multiple versions of the same function and switch based on the first parameter type. Wanted to get people's thoughts.

Reply:
Very very cool. I like this a lot more than taking a big `Union` of types and then having to handle them with a bunch of `if` statements at the top of the function. It's also nice because it encourages (a) static typing and (b) having one canonical version of function, and then one or more "aliases" which end up lightly wrapping the canonical verison.

Reply:
I like that it encourages static typing, but it seems to make the code longer and _slightly_ harder to follow compared to just having different names for the different versions of each function.

Reply:
Okay, a major knock against it is that it seems to deeply confuse Pylance, making it so that the auto-complete is useless. The more I read about singledispatch and its handling in various tools, the more it seems to fall in the category of "this is a weird thing that has to be special cased everywhere". I'm leaning towards actually making it "a big Union of types and then having to handle them with a bunch of if statements at the top of the function" at this point because of this. It's "ugly" but it's simple and tooling-friendly.